tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1591 @@
+		_, err := f.Read(length)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			f.Close()
+			return nil, err
+		}
+
+		dataLength := btou64(length)
+		if dataLength == 0 {
+			break
+		}
+
+		// read in the compressed block and decod it
+		b := make([]byte, dataLength)
+
+		_, err = f.Read(b)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			// print the error and move on since we can't recover the file
+			l.logger.Println("error reading lenght of metadata: ", err.Error())
+			break
+		}
+
+		buf, err := snappy.Decode(nil, b)
+		if err != nil {
+			// print the error and move on since we can't recover the file
+			l.logger.Println("error reading compressed metadata info: ", err.Error())
+			break
+		}
+
+		sf := &seriesAndFields{}
+		if err := json.Unmarshal(buf, sf); err != nil {
+			// print the error and move on since we can't recover the file
+			l.logger.Println("error unmarshaling json for new series and fields: ", err.Error())
+			break
+		}
+
+		a = append(a, sf)
+	}
+
+	if err := f.Close(); err != nil {
+		return nil, err
+	}
+
+	return a, nil
+}
+
+// writeSeriesAndFields will write the compressed fields and series to the meta file. This file persists the data
+// in case the server gets shutdown before the WAL has a chance to flush everything to the cache. By default this
+// file is flushed on start when bz1 calls LoadMetaDataIndex
+func (l *Log) writeSeriesAndFields(fields map[string]*tsdb.MeasurementFields, series []*tsdb.SeriesCreate) error {
+	if len(fields) == 0 && len(series) == 0 {
+		return nil
+	}
+
+	sf := &seriesAndFields{Fields: fields, Series: series}
+	b, err := json.Marshal(sf)
+	if err != nil {
+		return err
+	}
+	cb := snappy.Encode(nil, b)
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if _, err := l.metaFile.Write(u64tob(uint64(len(cb)))); err != nil {
+		return err
+	}
+
+	if _, err := l.metaFile.Write(cb); err != nil {
+		return err
+	}
+
+	return l.metaFile.Sync()
+}
+
+// nextMetaFile will close the current file if there is one open and open a new file to log
+// metadata updates to. This function assumes that you've locked l.mu elsewhere.
+func (l *Log) nextMetaFile() error {
+	if l.metaFile != nil {
+		if err := l.metaFile.Close(); err != nil {
+			return err
+		}
+	}
+
+	metaFiles, err := l.metadataFiles()
+	if err != nil {
+		return err
+	}
+
+	id := 0
+	if len(metaFiles) > 0 {
+		num := strings.Split(filepath.Base(metaFiles[len(metaFiles)-1]), ".")[0]
+		n, err := strconv.ParseInt(num, 10, 32)
+
+		if err != nil {
+			return err
+		}
+
+		id = int(n)
+	}
+
+	nextFileName := filepath.Join(l.path, fmt.Sprintf("%06d.%s", id, MetaFileExtension))
+	l.metaFile, err = os.OpenFile(nextFileName, os.O_CREATE|os.O_RDWR, 0666)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// metadataFiles returns the files in the WAL directory with the MetaFileExtension
+func (l *Log) metadataFiles() ([]string, error) {
+	path := filepath.Join(l.path, fmt.Sprintf("*.%s", MetaFileExtension))
```

Comment:
I'm not sure if Glob takes a full regex. Even so, I'm not really worried about it since this directory is only used for the WAL, and even then, no other files in our system use `.meta` in the name.